### ISSUE TITLE
Improved de-duping in pnpm

### ISSUE THREAD

**Issue by nickpape-msft** [pnpm/pnpm#952]
### pnpm version: `latest` (1.23.2)

pnpm has a vastly different (and more technically correct) linking and installation algorithm than npm does, but it appears to have a couple of downsides when it comes to de-duplication.

As npm installs dependencies, it looks at the folder structure on disk to decide if a satisfactory dependency has already been installed in the `node_modules` folder in the proper location (that is to say, in a parent folder of a dependency). For npm, this is very useful as it helps to reduce multiple installations of dependency, which was a large problem with `npm@2`.

However, due to pnpm's superior algorithm, at first this level of deduplication does not really seem necessary. In pnpm-land, the dependency is always installed only once (in the store), regardless of where it is used or depended on. However there are a couple of cases where having a deduplication story similar to `npm` would be useful.

Imagine that you are bundling production code into a single javascript file that will be deployed to your website. If you have several dependencies that are using a single UI library (for example, [office-ui-fabric-react](https://github.com/OfficeDev/office-ui-fabric-react)), they may all have different, but overlapping specifiers for office-ui-fabric-react. For example imagine your project `my-project` has direct dependencies on `lib-a`, `lib-b`, and `lib-c`. `lib-a` requires `>=1.0.0`, `lib-b` requires `1.11.1`, and `lib-c` requires `>=1.9.0 <2.0.0`. From our testing, it appears that pnpm always attempts to link to the latest compatible version of a dependency, which in this case may lead to `lib-a` being linked to `2.0.0`, `lib-b` being linked to `1.11.1` and `lib-c` being linked to `1.20.0`. In addition to being confusing to developers, when we create our production bundle, we may end up bundling components every single version of `office-ui-fabric-react` -- that means 3x copies of a button, 3x copies of a dropdown, 3x copies of a persona card, etc. However, the version `1.11.1` satisfies ALL the above version specifiers, and thus this problem could be avoided.

In npm-land, we can avoid this problem by adding a direct dependency to the `package.json`. npm *always* installs all direct dependencies first, before it begins installing indirect/second-order dependencies (which happens in a non-deterministic order).

Our build tool, [Rush](https://github.com/Microsoft/web-build-tools/wiki/Rush) is a mono-repo manager which creates a fake `package.json` file which is actually used for installations. This `package.json` links to several `.tgz` files for each individual project, allowing different projects to use different versions of a single dependency. It also has a feature called [pinned versions](https://github.com/Microsoft/web-build-tools/wiki/Rush-~-Pinned-versions) which is designed to let you control the top-level version of indirect dependencies by promoting that specific version into the top-level `package.json`. Additionally, in order to replicate the behavior of NPM, it has a feature called "implicitly pinned dependencies", which is a fancy phrase meaning that if the version specifier for a dependency is the same in every project, Rush will automatically pin it into the top-level `package.json`.

We also have an internal tool which automatically propagates packages across multiple repositories. Since we have several disjoint repositories which all use office-ui-fabric-react, they all have different version specifiers. The tool works by pulling the `package.json` information for all these "evergreen" packages, and choosing the newest version of those packages that results in no duplication (similar to how choosing `office-ui-fabric-react@1.11.1` in the above example satisfies all dependencies without causing any duplication). This tool works by utilizing the de-duped version with the "pinned versions" feature in Rush, which ensures that specific version is installed at the root, thereby causing NPM to re-use the de-duplicated version rather than installing multiple copies of the dependency.

We have been doing quite a bit of work to get Rush to work with pnpm (it is overall a much better utility than vanilla npm), however we have been running into issues with this. We have been kicking around a number of possible solutions, and came up with one that we thought would be interesting to propose. It would work something like this.

Imagine we are trying to figure out which version of `colors` to use. 

* When choosing which version of a dependency to install, crawl the dependency graph in reverse (i.e. back to the root).
  * If this package has already taken a dependency on `colors` that satisfies the version we are looking for, then re-use it.
* If nothing in the dependency graph has used `colors`, utilize the latest satisfactory version of colors that has already been installed anywhere in the repository.

This would help mitigate our problem, as we would place the de-duped version in the root `package.json`, causing it to be installed first, which would force everyone to use the de-duplicated version.

**Comment by pgonzal** [comment]
Maybe there could be a configuration switch where you chose between (1) favoring latest version of indirect dependencies, versus the above strategy of (2) favoring compatibility with NPM and minimizing bundle size.

I would argue that (2) should be the default behavior.  It will make adoption easier. People are very likely to give up if "pnpm install" produces a bad result the first time they run it.  It has advantages for NodeJS projects, but it's a pretty unworkable algorithm for the browser environment (where bundle size is crucial).

> Imagine we are trying to figure out which version of colors to use.
> 
> - When choosing which version of a dependency to install, crawl the dependency graph in reverse (i.e. back to the root).
>   - If this package has already taken a dependency on colors that satisfies the version we are looking for, then re-use it.
> - If nothing in the dependency graph has used colors, utilize the latest satisfactory version of colors that has already been installed anywhere in the repository.
> 

Consider this dependency chain (A is the project, D is the most indirect dependency):

```
- A asks for L version 1.2.3  --> gets 1.2.3
  - B asks for L version ^1.0.0   --> gets 1.2.3
    - C
      - D asks for L version ^2.0.0 --> gets 2.3.4
  - E asks for version 2.0.2 --> gets 2.0.2
```

So your basic principle would be that:

- B gets 1.2.3 because its parent asked for something compatible
- D gets version 2.0.3 because nobody in its parent chain asked for anything compatible

D does not get 2.0.2 because E is part of a separate chain of dependencies.

NPM 5 seems to do more global analysis of the tree than before, but I suspect it still basically works in this way.  Unlike pnpm, its intelligence is limited by the **node_modules** folder design, which greatly limits the version mappings it can do (without massive duplication of folders).

**Comment by zkochan** [comment]
Related #890 
cc @donaldpipowitch.

**Comment by zkochan** [comment]
This is hard or even impossible to do with pnpm w/o huge performance penalties.

One version of one package has always only one set of resolved dependencies in a node_modules created by pnpm. Lets' assume `a@1.0.0` has two dependencies: `b@2` and `c@2`. `a@1.0.0` can be in two different places in node_modules: `foo > a` and `bar > a`. During installation, pnpm will walk down the dependency graph to `a`, but it will resolve its dependencies only once. Either under `foo` or under `bar`. It will depend on which package's meta will be downloaded faster from the registry.

Now lets' assume `bar` also depends on `b@<2.5` and `foo` depends on `c@<2.5`.

Lets' assume the latest versions of both `b` and `c` is `2.6.0.

Now if we'll add this dedupe logic, `a` will get installed either 

`b@2.6.0` + `c@2.0.0`
or
`b@2.0.0` + `c@2.6.0`

* * *

An alternative solution might be to keep all the already resolved packages in some global state and prefer the resolved versions that are already there. So if `foo@2.5.0` is already used somewhere than use it for resolving `foo@^2.0.0`, even if 2.6.0 is available in the registry.

This solution would not slow down pnpm.

* * *

What are the real-world scenarios for this feature though? I understand the bundle filesize issue. However, packages should never be required using exact versioning or using some weird ranges like `>=2.0.0,<=2.6.0`. Why would a package break after a non-breaking change? If a package has specs like that, we can probably print a warning or something. That should not be considered normal

**Comment by pgonzal** [comment]
Suppose we use React as our view engine, and it has major versions 15 and 16.

App1 has upgraded to version 16. App2 is still stuck on 15 for the time being, because migrating is expensive, and sometimes has to wait for an upstream fix.

Both App1 and App2 use a library called React-Utils. The latest version of React-Utils is compatible with both versions 15 and 16 of React. It needs React, but it doesn't care which version, so it exorsswes the dependency as a SemVer pattern like ">=15.1.0 <17.0.0".

With NPM's strategy, this allows App1 and App2 to use the latest version of React-Utils (with all the fixes, without worrying about complicated version pairings). But with PNPM, neither App1 nor App2 will get the version that it wants. Instead, PNPM will always bind React-Utils to the highest compatible version  of React. Neither the App1 nor the App2 developer wants this.

As I understand it, your current recommendation is for them to work around this by coding a hook script. This isn't a very realistic approach.

We have hundreds of projects, and rely on libraries developed by partner teams in separate repos, who don't known the details of our dependencies. The guy who makes React-Utils doesn't work on App1 or App2, or want to know how they work. He just wants to state which versions of React he is compatible with. We have many intermediary libraries in our ecosystem that work like this.

**Comment by zkochan** [comment]
I see.

In this case, even Yarn works like pnpm. They have a special flag called [--flat](https://yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-flat). If we can't agree on a solution that keeps the current performance of pnpm, then the change will have to be under a flag

**Comment by pgonzal** [comment]
@nickpape-msft It might be useful to set up a minimal repro that demonstrates the problem.  Then we could compare the version bindings chosen by npm vs pnpm vs yarn.  And we could propose how the pnpm symlinks might be changed to solve the problem (without harming performance), or maybe we could find a different approach.  The problem is real, but the solution we're discussing seems a little vague right now.

**Comment by nickpape-msft** [comment]
@zkochan , @pgonzal and I discussed this issue more, and we think that the issue Pete was describing is a much more difficult problem than the one we are immediately trying to solve (which is much similar to the problem I described in the initial e-mail).

### Description

We sketched out a set of rules that the algorithm would follow as it installs dependencies. Assuming that the dependency tree is solved in a breadth-first manner, the rules for the algorithm would be as follows. As for terminology:

* The **current package** is the one that are currently working.
* The **child** is the dependency which we are concerned about selecting a version for.
* The **selector" is the semver selector describing which version of the **child** the **current package** is compatible with.
* The **parent** is the package which listed another package (the **current package**) in its dependency list.

0) If we have already installed this version of the **current package**, prefer to use it. E.g., if we have already fully installed `colors@1.0.0`, then we should not need to solve dependencies for, or create a second `colors@1.0.0`.

1) At this point we are picking a version for one of the **children**/dependencies of the current package. First, we traverse the dependency graph back to the root package, if any parent has a direct dependency on the child, we need to look at that version:
    * If it satisfies the **selector**, then we re-use the version that **parent** has already installed.
    * If it does not satisfy the **selector**, then we install the latest compatible version available from the registry.
    * If we cannot find a parent that has a direct dependency at all, then we install the latest compatible version from the registry.

This approach has several advantages:
* It should not be terribly expensive.
* It will reduce the total number of packages that are installed. This will speed installs.
* It will reduce the number of side-by-side packages, improving developer's lives.
* It behaves more similarly to npm. [However, whereas npm uses the current state of the node_modules folder, we will use the dependency graph].
* It allows developers to control indirect dependencies, perhaps to stop from upgrading to a buggy package (people don't always follow semver!).
* It solves our real world scenario of attempting to de-duplicate browser packages.

## Examples

### A trivial example
Imagine a developer is writing an app called `console-app`. He has a dependency on a `console-ui` library, which uses `colors`.
```
         console-app
         |
        [*]
         |
     console-ui (@1.0.0)
         |
     [^1.0.0]
         |
      colors (@1.9.0)
```

Now imagine this developer wants to develop with colors in his app. He adds a dependency on colors, but he decides to use `colors@1.1.0` (for whatever reason - maybe he is more comfortable with the older version, maybe he is already using that version in other apps, maybe the colors people changed "green" from dark to light, etc).

The current pnpm model will resolve this as follows:
```
         console-app
         |          \
        [*]          \
         |            \ 
         |             \   
console-ui (@1.0.0) colors (@1.1.0)
         |
     [^1.0.0]
         |
colors (@1.9.0)
```

There are some downsides to this. `console-ui` is using a different version of colors, so now things are being printed out in both light and dark green! We also have to install 2 copies of colors, which slows build times! 

Using the algorithm above, we would get the following tree:

```
         console-app
         |          \
        [*]          \
         |            \ 
         |             \   
console-ui (@1.0.0) colors (@1.1.0)
         |
     [^1.0.0]
         |
colors (@1.1.0)
```

This is great! Now we are using the same colors. But lo and behold, the maintainer of `console-ui` has only been testing against the latest version of `colors`, he doesn't realize that the latest `console-ui` no longer works with `colors@1.1.0`. In this case, the `console-app` developer will realize this, and bump his direct dependency on `colors` to something that will work for both of them.

[We have discussed more complex solutions in the future, involving installation "hints", which would allow a `package.json` to say whether itself or a dependency should be de-duplicated, or prefer the latest version. This behavior would allow `console-app` to let `console-ui` use the latest colors. It also solves a host of other problems, but we will outline this concept in another issue.]

### Example from my initial comment
The initial example I had is very straightforward:
```
             my-browser-app
            /      |      \
           /       |       \
          /        |        \     
        A@*       B@*       C@*
         |         |         |         
    >=1.0.0      1.2.3      ~1.2.0         
         |         |         |         
      lib        lib        lib
```

Currently, NPM might resolve this as follows:

```
             my-browser-app
            /      |      \
           /       |       \
          /        |        \     
        A@*       B@*       C@*
         |         |         |         
    >=1.0.0      1.2.3      ~1.2.0         
         |         |         |         
    lib@1.3.4   lib@1.2.3   lib@1.2.9
```

Now, when bundling the app, you are getting 3 versions of `lib` bundled. Sure, you might be able to fix this by writing a hook in your `pnpmfile.js`, but what if you are a browser library, and not an app, what about anybody who installs you? They will not have your complex `pnpmfile.js` that describes in detail what overrides are needed.

In npm land, you solve this by manually adding a direct dependency. For example, imagine you have a tool which calculates the versions of all libraries (direct and indirect) needed in order to stop any of them from being duplicated [we in fact have such a tool]. If we follow the above algorithm, we would end up in a state much similar to npm. Additionally, we have a single version of the package, speeding up installation dramatically!

```
             my-browser-app --------
            /      |      \         \
           *       *       *         1.2.3
          /        |        \         \
        A@1       B@1       C@1        lib@1.2.3
         |         |         |         
    >=1.0.0      1.2.3      ~1.2.0         
         |         |         |         
    lib@1.2.3   lib@1.2.3   lib@1.2.3
```

### A more complex example that covers every case

Imagine the following tree. [We don't care about version numbers as much in this example, as we are just trying to illustrate how the rules of the algorithm would work].

```
                             app
                              |
                +--------+----+----+--------+
                |        |         |        |          
               B@1       Q         D        F
                |        |      +--+----+
                |        |      |       |   
                F(bad)   F(bad) F(good) E
                                      |
                                     B@1
                                      |
                                      F
```

In this case, we would take the following steps:

1. We would install the latest versions of [**B**, **Q**, **D**, **F**]

2.
    * **D**'s **F** would resolve to the same version of **F** that the app is using, because the app is the first parent with a dependency on **F**, and it is compatible.
    * Both **B** and **Q** would get the latest version of **F** compatible with their specifiers, because the version the app is using is not compatible.
    * The latest version of **E** would be installed.

3. **E** depends on `B@1`. However, we have already solved a `B@1`, therefore, we will just reuse that installation. This means that this B will be getting the same `F (bad)` that we installed in step 2.

You can observe the following things about this tree:

1) We do not install the same version of a package multiple times with different dependencies.
2) We utilize the first parent's version of a dependency, but fall back to default behavior of it is not found or is compatible.
3) The version of F installed in an unrelated part of the tree does not affect other unrelated parts of the tree. E.g. the version that Q installs does not have an affect on anyone else. If this were not the case, we would have weird quantum entanglement and spooky action at a distance like NPM has (where someone somewhere in the tree downgrades a package and sets off a crazy chain reaction across the rest of the dependency graph).

@zkochan could you weigh in on the performance/usability impacts of a change like this?

**Comment by donaldpipowitch** [comment]
Just wanted to say thank you for this in-depth explanation. As far as I can tell that sounds like nice solution.

**Comment by zkochan** [comment]
@nickpape-msft what about this case?

Both B and D have E as dependency. B relies on the bad F and D relies on the good one, which F will be used for `B@1`? 
```
                             app
                              |
                +-------------+----+----+--------+
                |             |         |        |          
               B@1            Q         D        F
            +---+---+         |      +--+----+
            |       |         |      |       |   
           F(bad)   E      F(bad)    F(good) E
                    |                        |
                   B@1                      B@1
                    |                        |
                    F(good or bad?)          F(good or bad?)
```

**Comment by pgonzal** [comment]
Whatever we choose, it should probably be deterministic (even if it just goes in alphabetical order). NPM's [nondeterminism](https://docs.npmjs.com/how-npm-works/npm3-nondet) is annoying, since  if two people have the same package.json file and do their installs at the same time, they expect to end up in the same state.

**Comment by nickpape-msft** [comment]
@zkochan , that is a very good point. I'm not sure how we handle it. I agree with Pete that we need to come up with something deterministic. That is another attractive feature of pnpm for us.

**Comment by pgonzal** [comment]
PNPM could duplicate E and also B@1, the way peer dependencies are handled.  In some situations this might be the best solution.  But it should not be the default behavior, because it reintroduces the performance penalties @zkochan mentioned.  (Also, it can be counterintuitive:  If B@1 contains a class Button that inherits from a base class in F, then in the debugger you could find two different Button classes *with the same library version* but with different implementations/behavior. I bet most people assume that's impossible.)  When duplication is really needed, it could be explicitly requested via a custom package.json directive which we can invent later.

So, assuming we DON'T duplicate B@1, then the question is simply "which F wins?"  NPM's answer is "whoever got installed first," which is nondeterministic with incremental installs.  Some possible deterministic answers:

Which F wins?
1. Whichever F appears earliest in a breadth-first alphabetical tree traversal.
2. Whichever F has largest compatible version number

@zkochan do you have a preference?  If not, maybe we could just try something and evolve it over time?  These edge cases are interesting, but they aren't crucial for us.  Whereas the basic issue that Nick outlined is preventing us from adopting PNPM at all in our main repo.  We'd like to get that resolved quickly.

**Comment by zkochan** [comment]
We can try this approach if this is so crucial to you. If it won't be too heavy from performance point of view, then we can have it as the default behavior.

I may have time to do it this week but I cannot guarantee. If you'd like to do it, I think this part should be changed: https://github.com/pnpm/supi/blob/master/src/install/installMultiple.ts

**Comment by zkochan** [comment]
So I was looking into it today and I really don't see how it can be made deterministic w/o changes that would affect performance.

The only deterministic solution is to prefer packages that are the project's dependencies. I will implement this approach for now.

**Comment by pgonzal** [comment]
What is the perf problem? Is it during an incremental install (e.g. adding a new package to an existing node_modules folder), that you need to scan the entire tree again?

Or is it during a clean install, that you were trying to process HTTP responses as fast as they arrive, but now you need to wait for alphabetical order, so e.g. if A takes a long time then B, C, and D all get blocked?

**Comment by zkochan** [comment]
During a clean install. We have to wait till all dependencies are resolved on one level and only after that we can start resolving on the next level. Right now, pnpm can resolve concurrently, in any order, on different depths.

**Comment by pgonzal** [comment]
But we're only talking about fetching the package manifest, right? (The REST call that retrieves all the versions and their package.json data.) Not fetching or installing the tarball.

**Comment by zkochan** [comment]
We talk about everything. Delaying the package manifest fetch delays all the other things as well.

**Comment by zkochan** [comment]
As far as I understood from the initial post, in Rush you have the possibility to pin some versions of indirect dependencies. The way you do it, you add those dependencies as direct dependencies. With the change that I am making, your pinned dependencies will work with pnpm, same as they work with npm.

We can even support a more sophisticated approach. We can have a field that will contain "preferredVersions". Something like:

```json
"preferredVersions": {
    "library-a": "~1.0.0",
}
```

`library-a` will not be installed as a direct dependency but the specified range will influence the resolution of indirect dependencies.

**Comment by pgonzal** [comment]
Would it be much work to measure the performance cost and see if it is as significant as you assume?

I understand where you are coming from, however it seems really strange that if my registry is static, and I do two clean installs, that I could end up with two different version mappings due to network timings. I want my installs to be fast, but I don't want them to be a crapshoot. :-) In practice it probably is not very noticeable most of the time, but random algorithms are an absolute pain to diagnose when something goes wrong. (For example, we now use SQL hints to completely disable the query planner in production, because the speed gain wasn't worth the headache of dealing with a small percentage of bad query plans.)

Or maybe this could be a PNPM option?

**Comment by zkochan** [comment]
Please see my post above.

We can try it out but it will need a serious refactoring of `installMultiple.ts` first.

It is hard to imagine though, that it won't slow pnpm down. The simplest example would be a package with a really deep branch of dependencies. All the short branches would be resolved really fast, with good concurrency, then the long branch would be the last one, installing its dependencies with little concurrency. Currently, the speed of installation is **always maximally fast**, it doesn't depend on the structure of the dependency graph. Maybe that is why pnpm is sometimes 3 times faster than Yarn.

I do understand that sometimes speed is not the most important factor but in this case, I think the proposed solution is a good compromise. If in this case, pnpm would choose de-duplication over speed, there would be no need in pnpm, because it would become a shameless npm clone.

**Comment by zkochan** [comment]
I may sound too pessimistic and strict but pnpm's speed is indeed really fragile and on the edge. In August I've spent a whole month trying to refactor pnpm's installation engine. I used reactive programming and it allowed me basically to manipulate the order of different operations pnpm does during installation. (Here's the PR: https://github.com/pnpm/supi/pull/8).

After a whole month of experimenting, I was unable to come up with a rewrite that would be as performant as the current version. 

For context, [ied](https://github.com/alexanderGugel/ied/) uses reactive programming and is faster than pnpm. But ied does not support peer dependencies and it makes things a lot easier.

**Comment by pgonzal** [comment]
Thanks. I will think about it also. I have a suspicion that there may be some clever algorithm that would make it deterministic without slowing down too much, based on the suspicion that the cases where actual differences occur are uncommon. Like maybe you could fetch based on an optimistic assumption and backtrack in the rare case where it is wrong. Something like that...

**Comment by zkochan** [comment]
Published [1.24.0](https://github.com/pnpm/pnpm/releases/tag/v1.24.0), which has now partial deduplication. The version specs of the direct dependencies are taken into consideration when resolving the indirect dependencies. When possible, the version from the root are reused.

**Comment by pgonzal** [comment]
Thanks a ton! We on vacation this week, but we"ll definitely give it a try when we get back and let you know the result. Happy holidays!

**Comment by nickpape-msft** [comment]
Thanks for fixing this!
